### ISSUE TITLE
Support haddock response files

### ISSUE THREAD

**Issue by srayuws** [haskell/cabal#1681]
Enviornment: 
windows 7 64bit , 
haskell-platform 2013.2, 
cabal.exe  version 1.18.0.2 using version 1.18.1.2 of the Cabal library
haddock.exe  version 2.13.2
 
HOW I MEET THIS:
I have a fresh haskell-platform 2013.2, and run the following:
```
# edit the config file to enable documentation
cabal update
cabal install cabal-install
cabal update
cabal install yesod  # this will install a lot hackages and run for minutes
```
At the end of the output, there are a few lines:
```
Updating documentation index
C:\Users\srayuws\AppData\Roaming\cabal\doc\index.html
cabal: C:\Program Files (x86)\Haskell Platform\2013.2.0.0\bin\haddock.exe:
does not exist
```
At this time, if I install or reinstall any hackage, it will say haddock.exe does not exist. 
By the way, it seems only the last installed hackage could not find haddock.exe, the document for its dependencies are gererated well.

But if i run ```"C:\Program Files (x86)\Haskell Platform\2013.2.0.0\bin\haddock.exe" --version ```, it works fine.

I have also noticed that after delete the ```doc``` directory (with about 100+ hackages in it) in ```%APPDATA%\cabal\```, then reinstall some hackages which could not find haddock.exe before, everything goes fine again...

**Comment by randen** [comment]
In building the Haskell Platform 2015.2.0.0 (for GHC 7.10.1) on the Windows platform, we are now unable to build the haddocks for OpenGLRaw, due to exactly the problem noted in haskell/haddock#285 so adding another vote and some urgency for this feature in Cabal.

One question I have is how cabal-install will expose this at the command-line level since cabal-install's options are a mix of its own and haddock's, so would the user specify only some of the haddock options in the response file?  Or would Cabal re-arrange the user-provided response file contents to include some parts that it adds to the actual haddock invocation?  Or would cabal-install allow its own response file for all of its options (including the ones it passes over to any other executable), then, perhaps optionally (e.g., looking at the version number of, in this case, haddock) supply the invoked executable with a response file of Cabal's own creation?  Could get complicated?

The simple approach I took in my experiments to fix this was to not do a full-on "response file" approach (i.e., *all* command line arguments in one file) but only to allow specifying the file list to haddock via a file, which eventually would still hit the limitation if using many other non-file options (e.g., --read-interface lists can be very long as well), but it did avoid some of the things to consider I mention above.  Just wanted to mention in case it helps you.

In any event, the solution in Cabal and cabal-install will need to match what haddock does, and then the HP build code will need to invoke cabal-install to match.

**Comment by Fuuzetsu** [comment]
https://github.com/haskell/haddock/commit/d567a12b2d24bab610cd7e8f8014d40c7615e24d

**Comment by randen** [comment]
What does a milestone of "bottom" indicate?  The problem we have building the docs for large libraries on the Windows platform needs some kind of solution.  If using response files is not a good solution for some reason (understandably, it is not needed for other environments), but we do need something or Windows builds will have missing docs for large libraries (and may even flat out fail to complete).

The Haskell Platform team humbly requests that a fix for this be included in the haddock which is shipped whatever the next official ghc release, likely 7.10.2.

**Comment by Fuuzetsu** [comment]
There is a problem with the Haddock's implementation of response files which probably needs to be fixed first.

**Comment by 23Skidoo** [comment]
> There is a problem with the Haddock's implementation of response files which probably needs to be fixed first.

That is, https://github.com/haskell/haddock/issues/379, without which response files are pretty useless.

**Comment by Fuuzetsu** [comment]
Right. I actually realised that we could have just one generic implementation in `Cabal` and Haddock could import that and feed args through it before doing Haddock-y stuff. It'd ensure consistent implementation which I believe is what we're after.

**Comment by 23Skidoo** [comment]
Can Haddock depend on Cabal?

**Comment by Fuuzetsu** [comment]
It already does.

https://github.com/haskell/haddock/blob/master/haddock.cabal#L59 and a similar line in `haddock-api` if we're not in GHC tree. This is OK because `Cabal` doesn't depend on Haddock and comes with GHC.

**Comment by 23Skidoo** [comment]
`process` or `base` are probably more appropriate places for this - I bet that `getArgsWithResponseFiles` could be generally useful.

**Comment by NightRa** [comment]
Any progress on this?

**Comment by 23Skidoo** [comment]
Fixed by #2746. 

@NightRa Thanks for the heads-up.
